src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/RazorDynamicFileInfoProvider.cs

```diff
@@ -101,15 +122,42 @@ public Task RemoveDynamicFileInfoAsync(ProjectId projectId, string? projectFileP
             RemoveRazorDynamicFileInfoMethodName, notificationParams, cancellationToken).AsTask();
     }
 
-    private sealed class EmptyStringTextLoader : TextLoader
+    private ValueTask UpdateAsync(ImmutableSegmentedList<string> paths, CancellationToken token)
     {
-        public static readonly TextLoader Instance = new EmptyStringTextLoader();
+        foreach (var path in paths)
+        {
+            token.ThrowIfCancellationRequested();
+            Updated?.Invoke(this, path);
+        }
+
+        return ValueTask.CompletedTask;
+    }
 
-        private EmptyStringTextLoader() { }
+    private sealed class EmptyStringTextLoader() : TextLoader
+    {
+        public static readonly TextLoader Instance = new EmptyStringTextLoader();
 
         public override Task<TextAndVersion> LoadTextAndVersionAsync(LoadTextOptions options, CancellationToken cancellationToken)
         {
             return Task.FromResult(TextAndVersion.Create(SourceText.From(""), VersionStamp.Default));
         }
     }
+
+    private sealed class TextChangesTextLoader(Document? document, IEnumerable<TextChange> changes) : TextLoader
+    {
+        public override async Task<TextAndVersion> LoadTextAndVersionAsync(LoadTextOptions options, CancellationToken cancellationToken)
+        {
+            var sourceText = document is null
```

Comment:
If document is null, but changes isn't, then aren't we dropping edits on the floor? 

For a new document, it makes sense that the document would be null until after the `GetDynamicFileInfo` method has finished executing, but doesn't that mean that when this method is called, its not null any more? Perhaps the look up of the document should be in this method, not when the request is made?

Reply:
I guess if document is null we should be able to assume changes is just the whole document. We could add a check for that? What edits would be dropped without that? 

This is definitely only intended for a new document so anything that can better reflect that is good.

Reply:
Oh whoops, yes, I missed that if document was null its just as though we're starting with a blank document. That makes complete sense, I retract my comment.